src/crypto/crypto_keys.cc

```diff
@@ -1134,6 +1136,51 @@ void KeyObjectHandle::InitEDRaw(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(true);
 }
 
+void KeyObjectHandle::Equals(const FunctionCallbackInfo<Value>& args) {
+  std::shared_ptr<KeyObjectData> key =
+    Unwrap<KeyObjectHandle>(args.Holder())->Data();
+  std::shared_ptr<KeyObjectData> key2 =
+    Unwrap<KeyObjectHandle>(args[0].As<Object>())->Data();
+
+  KeyType keyType = key->GetKeyType();
+  CHECK_EQ(keyType, key2->GetKeyType());
+
+  bool ret;
+  switch (keyType) {
+    case kKeyTypeSecret: {
+      size_t size = key->GetSymmetricKeySize();
+      if (size == key2->GetSymmetricKeySize())
+        ret = CRYPTO_memcmp(
+          key->GetSymmetricKey(),
+          key2->GetSymmetricKey(),
+          size) == 0;
+      else
+        ret = false;
+      break;
+    }
+    case kKeyTypePublic:
+      // Fall through
+    case kKeyTypePrivate: {
+      EVP_PKEY* pkey = key->GetAsymmetricKey().get();
+      EVP_PKEY* pkey2 = key2->GetAsymmetricKey().get();
+      if (EVP_PKEY_id(pkey) == EVP_PKEY_id(pkey2)) {
+#if OPENSSL_VERSION_MAJOR >= 3
+          ret = EVP_PKEY_eq(pkey, pkey2) == 1;
+#else
+          ret = EVP_PKEY_cmp(pkey, pkey2) == 1;
```

Comment:
Based on the OpenSSL documentation, this will not behave in the same way as `EVP_PKEY_eq`.

Reply:
In 3.x `EVP_PKEY_cmp` is an alias for `EVP_PKEY_eq`. In 3.x `EVP_PKEY_cmp` is deprecated. In 1.1.1 `EVP_PKEY_eq` does not exist.

Note from the docs on the difference is as follows:

> The function EVP_PKEY_cmp() previously only checked the key parameters (if there are any) and the public key, assuming that there always was a public key and that private key equality could be derived from that. Because it's no longer assumed that the private key in an [EVP_PKEY(3)](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY.html) is always accompanied by a public key, the comparison can not rely on public key comparison alone.

Furthermore, because of the above `EVP_PKEY_eq` does not have the affordance to check a public with a private key.

Our use of both `EVP_PKEY_cmp` and `EVP_PKEY_eq` never hits the above use case since we make sure the key types are the same.

Since `EVP_PKEY_cmp` is just an alias for `EVP_PKEY_eq` in 3.x I believe the assumption that there's always a public key accompanying a private key doesn't hold true only for keys introduced (or to be introduced) in 3.x

Because of the above I don't believe the difference in behaviour affects our use of these functions.

Reply:
https://github.com/openssl/openssl/issues/16267

Reply:
Alright, thanks for the write-up!